policy.go

```diff
@@ -399,218 +393,252 @@ func createPcrPolicyCounter(tpm *tpm2.TPMContext, handle tpm2.Handle, updateKeyN
+// that the NV indices have names that are common to all devices and can always be recreated, but they have authorization policies
+// that enforce a creation sequence that means that they can only be created in the locked state. The way this works is as follows:
+// - A bootstrap index is created at handle 2 with TPMA_NV_AUTHREAD and TPMA_NV_AUTHWRITE attributes and an empty authorization
+//   policy.
+// - An empty write is performed to initialize the bootstrap index.
+// - An index is created at handle 1 with TPMA_NV_AUTHREAD, TPMA_NV_POLICY_WRITE and TPMA_NV_READ_STCLEAR attributes, and an
+//   authorization policy that can only be satisfied by the presence of the initialized bootstrap index.
+// - An empty write is performed to initialize index 1 using a policy session.
+// - The bootstrap index is undefined and another index is created at handle 2 with TPMA_NV_AUTHREAD and TPMA_NV_POLICY_WRITE
+//   attributes, and an authorization policy that can only be satisfied by the presence of the initialized index at handle 1 with
+//   its read lock enabled.
+// - The read lock for the index at handle 1 is enabled.
+// - An empty write is performed to initialize index 2 using a policy session.
+//
+// Authorization policies for sealed keys must assert the presence of both of these NV indices using the names of the initialized
+// indices without the read lock enabled. Enabling the read lock on index 1 by calling LockAccessToSealedKeys will disable access to
+// these keys.
+//
+// It isn't possible to recreate these indices in their unlocked state. Although they can be recovered if they are undefined
+// accidentally, they will not return to their unlocked state until the next TPM reset or restart.
+//
+// An adversary can't undefine and redefine either index in order to reenable access to sealed keys. Eg, say an adversary performs the
+// following actions:
+// - They undefine and redefine index 1. The new index isn't read locked, but sealed keys can't be unsealed until the new index has
+//   been written to. But it has an authorization policy that requires index 2 to be undefined and replaced with the bootstrap index.
+// - They undefine index 2 and create the boostrap index in its place.
+// - They perform an empty write to index 1 to initialize it. Index 1 now has the expected name, but sealed keys still can't be
+//   accessed because index 2 no longer exists.
+// - They undefine the bootstrap index again and redefine index 2. However, sealed keys still can't be unsealed until the new index
+//   has been written to. But it has an authorization policy that requires the read lock to be enabled on index 1.
+// - They enable the read lock on index 1.
+// - They perform an empty write to index 2 to initialize it. Index 2 now has the expected name, but sealed keys still can't be
+//   accessed because index 1 is read locked. This can ony be undone by performing a TPM reset or restart.
```

Comment:
is there are a test that matches this hypothetical attack? also this explanation should probably emphasize names even more.  I start to be a bit worried about these long explanation before functions that are kind of then hard to match up with code because the code is sometimes spread over helpers in sometimes asymmetric ways.

Reply:
I've added a test case for this.